Non axis aligned cutting tool in sliceviewer

RichardWaiteSTFC opened the pull request:
**Description of work.**

Added new tool for non axis aligned cuts - see roadmap link in issue for more detailed requirements.

**To test:**
(0) Repeat the instructions in PR #33362 - (there was a  regression of this bug fix that has been re-implemented in this PR)
(1) Check the cutting tool button is only enabled for 3D MD workspaces where all dimensions are Q by opening the following workspaces in sliceviewer. It should only be enabled for the `ws_3D` and `ws_3D_QLab` workspaces (see comment for details) - the first 3 column headers for the vectors should be a*,b*,c* and Qx,Qy,Qz for the two workspaces respectively.

![image](https://user-images.githubusercontent.com/55979119/165113746-37cc3fed-2acb-4d7e-960b-120d67f0d316.png)
```
Load(Filename='CNCS_7860_event.nxs', OutputWorkspace='CNCS_7860_event')  # disabled (MatrixWorkspace)

ws_2D = CreateMDWorkspace(Dimensions='2', Extents='-5,5,-4,4', Names='H,K',
                          Units='r.l.u.,r.l.u.', Frames='HKL,HKL',
                          SplitInto='2', SplitThreshold='50')  #  disabled (2D MD)

ws_3D = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                          Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                          Frames='HKL,HKL,HKL', SplitInto='2', SplitThreshold='50')  # enabled!
    
ws_3D_nonQdim = CreateMDWorkspace(Dimensions=3, Extents=[-1, 1, -1, 1, -1, 1], 
                          Names="E,H,K", Frames='General Frame,HKL,HKL', 
                          Units='meV,r.l.u.,r.l.u.')  # disabled (3D but 1 non-Q)
                          
ws_4D = CreateMDWorkspace(Dimensions=4, Extents=[-1, 1, -1, 1, -1, 1, -1, 1], 
                          Names="E,H,K,L", Frames='General Frame,HKL,HKL,HKL', 
                          Units='meV,r.l.u.,r.l.u.,r.l.u.')  # disabled (4D - one non-Q)

ws_3D_QLab = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                               Names='Q_lab_x,Q_lab_y,Q_lab_z', Units='U,U,U',
                               Frames='QLab,QLab,QLab', SplitInto='2', SplitThreshold='50')  # enabled!
```
(2) Close any sliceviewer windows and clear the workspaces in the ADS
(3) Run the following and open `ws` in sliceviewer.
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                          Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                          Frames='HKL,HKL,HKL', SplitInto='2', SplitThreshold='50') 
expt_info = CreateSampleWorkspace()
SetUB(expt_info, 1,1,2,90,90,120)
ws.addExperimentInfo(expt_info)
# make some fake data
FakeMDEventData(ws, UniformParams='1e5', PeakParams='1e+05,0,0,1,0.3', RandomSeed='3873875')
```
(4) Click on non-ortho view of H and K axes - this should disable the cutting tool button
(5) Turn off non-ortho view and check that opening the cut tool:
- Disables non-ortho view
- Disables ROI tool
- Disables line plots

It should look like this
![image](https://user-images.githubusercontent.com/55979119/165118214-39255136-6a2e-4b08-a219-b3bb22f73722.png)

(6) Check that transposing the axes (X <-> Y) will swap the u1 and u2 vectors in the table
(7) Set the axes to (X,Y) = (L, K) - check u1 = [0,0,1] and u2 = [0,1,0]
(8) Change the slice point of H to be 0 - it should look like this
![image](https://user-images.githubusercontent.com/55979119/165118671-58c5aead-c527-484b-969d-db07e303b1cb.png)
(9) In the table double the step along u1 (i.e. set step to 0.12) - this should change nbins = 25 along u1
(10) Set the nbins = 50 along u1 - the step should go back to it's original value (0.06)
(11) Set the stop for u1 to 0 and check that 
- step size = 0.03
- the cut representation line on the colorfill plot has the correct start/stop
- the 1D plot in the cut viewer pane has the correct axes limits

(12) Set step of u1 to be 2 (i.e. greater than the extent of the cut) - this should set nbins=1 and step = 1.5 and put the cut along u2 with nbins = 50.
(13) transpose the axes so now (X,Y) = (K,L) - the cut should have 50 bins along K (default value)
(14) Change the nbins of u2 to 50 (it should set nbins=1 for u1 and change the step=4). Check the white line of the cut representation on the colorfill plot is now vertical.
(15) Try to change the a* column of the u1 to 1 (this would take u1 out of the plane of the slice, i.e. not orthogonal to u3) - it should reset to 0 - i.e. u1 = [0,1,0].
(16) Click and hold down on one of the red markers with white face on the colorfill plot and drag, release at ~K=1.
- This should reset the vectors int he table such that the cut is along u1 = [0,0,1] - i.e. u1 <-> u2
- The thickness along u2 should be adjusted to ~2

(17) Set the step of u2 = 2 in the table, check that it sets (start,stop) = (-1,1)
(18) Drag the top white marker of the cut representation up to L~2
- u1 ~ [0,0,-1]  and u2 ~ [0,-1,0]
- There should be a peak in the 1D plot at x~-1

(19) For u1 change c* = -1 and b* = 0 - check that u2 = [0,-1,0]
(20) To change the centre of the cut move the central white marker of the cut representation to (K,L) ~ (2,0),
- The entire cut representation should move
- The axes label of the 1D plot should be ~`(0.0, 2.0,0.0-1.0x) in 3.14 Ang^-1`
- There should be no peak on the 1D plot

(21) Increase the thickness by dragging the left red marker of the cut representation to encompass the peak in the data - check the peak appears in the 1D plot at the right thickness.
(22) Play around with the direction of the cut by dragging the white markers at the end points of the white line - the vectors u1 and u3 should be orthogonal unit vectors.
(24) Reset the cut by transposing the axes so (X,Y) = (L,K)
(25) Double the slice thickness along H from 0.1 -> 0.2 (the counts of the peak in the 1D plot should double from ~2000 -> ~4000)
(26) Change c* of u1 from 1 -> -1
- The peak in the 1D plot should move from x= 1 -> -1
- Check u2 = [0,-1,0]


Fixes #33631
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

RichardWaiteSTFC committed:
Add cutting widget view to sliceviewer

RichardWaiteSTFC committed:
Fill cut viewer table with projection steps and bin params

Default cut will be along the x-axis, in the center of the y-axis with a width equal to 2 bin widths

RichardWaiteSTFC committed:
Get correct initial data limits in non orthog view when init table

Also moved some methods that access properties of dataview to base_presenter class

RichardWaiteSTFC committed:
Update table on dimension and slicepoint changing

RichardWaiteSTFC committed:
Perform and plot cut, add logic to change vects/bins on cell change

RichardWaiteSTFC committed:
Clear 1D plot axes to fix autoscaling

RichardWaiteSTFC committed:
Add horace style axis label to plot

RichardWaiteSTFC committed:
Reset negative nbins and steps

RichardWaiteSTFC committed:
Rely on ADS replace in plot and fix axes lims autoscaling on replace

RichardWaiteSTFC committed:
Make and plot initial cut on opening cut viewer

RichardWaiteSTFC committed:
Plot interactive cut representation on 2D colorfill

RichardWaiteSTFC committed:
Set wsname of 1D cut in sliceviewer model

RichardWaiteSTFC committed:
Fix bug in setting bin params and slicewidth when axes transposed

RichardWaiteSTFC committed:
Fix sign error in cut thickness and force start to be pt with min x

RichardWaiteSTFC committed:
Set default cut using axes limits only - not data extent

RichardWaiteSTFC committed:
Reformat xlabel of 1D plot - BinMD does not allow desired format

RichardWaiteSTFC committed:
Adjust extents when step of integrated dim changed in table

Also fixed bug that adjusts extents to ensure nbins is an integer for chosen step size along cut dimension

RichardWaiteSTFC committed:
Check user supplied vector in plane of slice otherwise reset

RichardWaiteSTFC committed:
disable lineplot and region select on opening cutviewer

RichardWaiteSTFC committed:
Fix regression of bug #33104 reveresed in PR #33639

Bug relates to support for dynamic rebinning when MDhisto workspace has been altered (e.g. scaled).
It was initially fixed in #33362 which introduced a function to determine whether the orignal
MDE workspace should be rebinned (it shouldn't if the MDHisto has been altered). In PR #33639
the function that checks whether a ws can be dynamically rebinned was moved and it no longer
checked whether the MDHisto had been altered (but the function to do so remained in the model).
This commit moves the function to check if the MDHisto had been altered to the WorkspaceInfo model
and fixes the regression.

RichardWaiteSTFC committed:
Disable the cut viewer if ws does not support dynamic rebinning

At present only enabled for 3D MD workspaces with 3 Q dims - this is temporary

RichardWaiteSTFC committed:
Put test back in for bug #33104 (deleted in PR #33639)

RichardWaiteSTFC committed:
Make cut view in cut presenter rather than sliceviewer view

All communication now between sliceviewer and cutviewer presenters, the data_view of sliceviewer no longer needs to hold the view as an attribute.

RichardWaiteSTFC committed:
Implement MVP for cutviewer widget

RichardWaiteSTFC committed:
Fix bug identifying cut axis when all nbins are 1

RichardWaiteSTFC committed:
Fix bug in cut plot xlabel formatting

RichardWaiteSTFC committed:
Turn off cursor tracking when cutting tool enabled

Varying precision on cirsor position causes plot to resize which interferes with interactive cutting tool

RichardWaiteSTFC committed:
Fix workspace info tets now have additional check for altered MDH

RichardWaiteSTFC committed:
Fix bug settiung the initial default number of bins in cut

RichardWaiteSTFC committed:
Better toolbar icon for cut view

RichardWaiteSTFC committed:
Add test file for cut viewer model

RichardWaiteSTFC committed:
Change Q dim column headers depending on frame

RichardWaiteSTFC committed:
Remove unnecessary atrribute in cut view model storing presenter

RichardWaiteSTFC committed:
Add test file for cut viewer presenter

RichardWaiteSTFC committed:
Disable cut viewer when non ortho view enabled

RichardWaiteSTFC committed:
Fix ValueError getting proj_matrix when MDE workspace has no exp info

RichardWaiteSTFC committed:
Add manual testing instructions for cut viewer

RichardWaiteSTFC committed:
Add cut viewer overview to slice viewer user docs

RichardWaiteSTFC committed:
Add test for non axis cut toggled, bug fix enabling ROI etc. buttons

Bug fix: now ROI, line plots, and if applicable non ortho view buttons enabled on closing cut viewer

RichardWaiteSTFC committed:
Add tests for when to support non axis aligned cuts

zjmorgan commented:
This works well for the 3D MD event workspaces and can even support the overlay of peaks simultaneously:
![image](https://user-images.githubusercontent.com/13754794/165383132-988c25d7-28ee-4035-8740-9fe7909b21df.png)

DannyHindson commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py:
```diff
@@ -0,0 +1,317 @@
+        xmin, xmax, ymin, ymax = self.get_start_end_points()
+        return 0.5 * (xmin + xmax), 0.5 * (ymin + ymax)
+
+    def get_perp_dir(self):
```
Please can you add some tests for some of the functions on this class (get_perp_dir\get_mid_point, get_start_end_points)

DannyHindson commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py:
```diff
@@ -0,0 +1,317 @@
+            textobj.set_fontsize(8)
+        self.figure.tight_layout()
+
+    def _format_cut_xabel(self):
```
should this be `format_cut_xlabel`?

DannyHindson commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py:
```diff
@@ -0,0 +1,317 @@
+        self.figure.axes[0].set_xlabel(xlab)
+
+
+class CutRepresentation:
```
I think this would ideally be in its own unit (similar to the peak representations in the peak viewer) - but see how much time you've got

DannyHindson reviewed (changes requested):
Good job on this. It's looking good and hopefully will be fine to get merged by the end of this week. I've added a small number of code review comments. Also I went through the testing instructions (v thorough!) and found a few things:
1) When I open the cutting tool, the axis label on the vertical axis is partially hidden. Is it possible to set the new windows size so that this is shown in full?
![image](https://user-images.githubusercontent.com/56295817/166469767-d5fb38d0-bf56-40d9-a36d-55e8d0455385.png)

2) If I set stop to be less than start on the u1 row in the table, the x axis on the plot doesn't update (but the line does)
3) When doing step 12 the plot does update. The red box gets taller
4) Setting step column to zero causes an error
5) If I open the slice viewer on the ws workspace from your script on the PR (step 3) and just toggle the new tool on and off I get a red "Singular matrix" error in the log:
![image](https://user-images.githubusercontent.com/56295817/166469943-8cffae51-69d7-44fe-85eb-71021cd875c6.png)

RichardWaiteSTFC committed:
Resize colorfill plot on splitter moving and widget added

The tight_layout is only applied if line plots ar not enabled - the thight_layout seems to ignore the line plot axes (fixing this is beyond the scope of this PR)

RichardWaiteSTFC committed:
Set min < max in extents to avoid negative step and ignore zero step

RichardWaiteSTFC committed:
Set nbins to 1 and swap cut direction if step greater than extent

Previously allowed all nbins==1 but then 1D cut would not be updated

RichardWaiteSTFC committed:
Move CutRepresentation class to separate file

RichardWaiteSTFC committed:
Separate out logic for validation of different bin params

Removed reference to column index in model and updated tests

RichardWaiteSTFC committed:
Add test file for cut representation

RichardWaiteSTFC committed:
Stop Singluar matrix error from tight_layout

RichardWaiteSTFC committed:
Fix typos in testing docs

RichardWaiteSTFC commented:
Thanks @DannyHindson for the feedback, I have put the cut representation in a separate file and added tests. There are a few changes in behaviour: when a user sets step =0 this is ignored (original value kept), when a user sets a stop < start the y are swapped, when a user sets step > extents then the nbins is set to 1 (as before) but then the cut is taken to be along the orthogonal vector in the plane with defualt nbins = 50 (so always have one direction with nbins > 1). The Singular matrix error came from pyplot tight_layout, it looks it might be a regression as the link I put in a code comment says it was fixed in v2.1) , I have 'fixed' it by checking for 0 figure height and width- ideally I would stop the figure height and width being set to 0 in the first place. I can have a look, but in the interests of time I thought this might be sufficient. Also I now call tight_layout on the colorfill plot when the slider is added or moved (so this should fixe the issue you saw for both the cut viewer and peak viewer).

RichardWaiteSTFC committed:
Resize cut viewer plot on splitter or window resizing

RichardWaiteSTFC committed:
Add release notes

DannyHindson reviewed (approved):
Good job! All looks good now, thanks for getting this done in the last couple of sprints so it could get into the release

RichardWaiteSTFC commented:
@mantidproject/gatekeepers - could someone please have a look at this, given this is the last day before the code freeze! Thanks!

gemmaguest commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py:
```diff
@@ -0,0 +1,210 @@
+    """Displays a table view of the PeaksWorkspace along with controls
+    to interact with the peaks.
+    """
+    def __init__(self, presenter, canvas, frame, parent=None):
```
Is there a way we could avoid leaking knowledge of the presenter to the view?

gemmaguest commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py:
```diff
@@ -0,0 +1,210 @@
+        self.frame = frame
+        self._setup_ui()
+        self._init_slice_table()
+        self.table.cellChanged.connect(self.on_cell_changed)
```
Could this signal connect to a slot on the presenter?

gemmaguest commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py:
```diff
@@ -0,0 +1,210 @@
+    # signals
+
+    def on_cell_changed(self, irow, icol):
+        self.set_bin_params(*self.presenter.validate_bin_params(irow, icol))
```
We're leaking knowledge of the presenter here - if this slot was on the presenter this could be avoided?

RichardWaiteSTFC replied:
Yes I should have made a slot for this, there are still other instances in the view where I call a method on the presenter as I didn't think this too bad, but I can try and remove them.

martyngigg reviewed (approved):
This looks great and I'm really glad we're going to get this in.

Can I ask that after I merge this we look at updating the code to use Python typing statements? I'm trying to encourage their use to improve the readability of Python code.

gemmaguest commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py:
```diff
@@ -0,0 +1,210 @@
+        return float(self.table.item(irow, 6).text())
+
+    def get_bin_params(self):
+        vectors = zeros((3, 3), dtype=float)
```
I'm a bit confused what all the hard-coded values mean in here. Is it worth adding static variables for e.g. what the column numbers mean (start=3, stop=4, etc.) and maybe what these values in the constructors mean (num-vectors, num-columns?).

There's quite a lot of logic here. Could it be moved to the presenter so it can be tested?

RichardWaiteSTFC replied:
See comment below for vectors always having 3 elements - in maintenance I will make a static variable called e.g. `NVEC_ELEM = 3` (or better). Perhaps the mapping of column indices to start/stop would be better done with an enum?

gemmaguest commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/cutviewer/view.py:
```diff
@@ -0,0 +1,210 @@
+        extents = zeros((2, 3), dtype=float)
+        nbins = zeros(3, dtype=int)
+        for ivec in range(vectors.shape[0]):
+            for icol in range(vectors.shape[0]):
```
I'm confused why this is iterating through `shape[0]` twice - should the second one be `shape[1]`?

RichardWaiteSTFC replied:
I could have used `shape[1]`, but in this case it doesn't matter as `vectors` is a 3x3 matrix and always will be even for 2D and 4D workspaces - we will always have two vectors in the plane of the slice and one out of the plane of the slice.

gemmaguest reviewed (commented):
Ah @martyngigg got there before me but I have a few comments that might be worth looking at in maintenance, mostly related to logic in the view that it might be possible to move to the presenter so we can test it. Feel free to push back if there was a reason for doing it the way you did (I'm not an expert in python GUIs!).

One thing I'm concerned about is I'm not sure it's iterating through the correct range in `get_bin_params` (although both ranges are `3` so we probably get away with this for now!).

martyngigg commented:
Sorry for heading in here quickly. Agreed that it would be good to attend to all of these things during maintenance.
